File: app-java/src/main/java/com/TritiumGaming/phasmophobiaevidencepicker/activities/mainmenus/marketplace/views/ThemeBundleCardView.java
package com.TritiumGaming.phasmophobiaevidencepicker.activities.mainmenus.marketplace.views;

import android.content.Context;
import android.util.AttributeSet;
import android.view.ContextThemeWrapper;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatTextView;

import com.TritiumGaming.phasmophobiaevidencepicker.R;
import com.TritiumGaming.phasmophobiaevidencepicker.data.models.settings.ThemeModel;
import com.TritiumGaming.phasmophobiaevidencepicker.utils.ColorUtils;
import com.TritiumGaming.phasmophobiaevidencepicker.data.models.firestore.theme.bundle.MarketThemeBundleModel;
import com.google.android.material.card.MaterialCardView;

public class ThemeBundleCardView extends MaterialCardView {

    @Nullable
    private MarketThemeBundleModel bundle = null;

    public ThemeBundleCardView(@NonNull Context context) {
        super(context, null);
    }

    public ThemeBundleCardView(@NonNull Context context, AttributeSet attrs) {
        super(context, attrs);

        init(context);
    }

    public ThemeBundleCardView(@NonNull Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        init(context);
    }

    public void init(@Nullable Context context) {
        inflate(context, R.layout.item_marketplace_bundle, this);

        setLayoutParams(
                new LinearLayout.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT));

        int strokeColor =
                ColorUtils.getColorFromAttribute(getContext(), R.attr.backgroundColorOnBackground);

        setRadius(16);

        setStrokeColor(ColorUtils.interpolate(
                getResources().getColor(R.color.white),
                strokeColor,
                .25f));
        setStrokeWidth(3);

        setUseCompatPadding(true);
        setClipToPadding(false);
    }

    public long getCreditCost() {
        if(bundle == null) {
            return 0;
        }

        return bundle.getDiscountedBuyCredits();
    }

    public void setBundle(@Nullable MarketThemeBundleModel bundle) {
        this.bundle = bundle;

        if(bundle == null) {
            return;
        }

        AppCompatTextView titleView = findViewById(R.id.label_bundleTitle);
        if(titleView != null) {
            titleView.setText(this.bundle.getName());
        }

        AppCompatTextView costView = findViewById(R.id.label_credits_cost);
        if(costView != null) {
            //costView.setText(String.valueOf(this.bundle.getBuyCredits()));
            costView.setText(String.valueOf(this.bundle.getDiscountedBuyCredits()));
        }

        buildThemes();
    }

    public void buildThemes() {
        LinearLayout themesList = findViewById(R.id.themesList);

        if(bundle != null) {
            for (ThemeModel theme : bundle.getThemes()) {

                ThemeBundleImageView card = new ThemeBundleImageView(
                        new ContextThemeWrapper(
                                getContext(),
                                theme.getStyle()),
                        null,
                        theme.getStyle());
                card.setTheme(theme);

                card.setLayoutParams(
                        new LinearLayout.LayoutParams(
                                ViewGroup.LayoutParams.MATCH_PARENT,
                                ViewGroup.LayoutParams.MATCH_PARENT, 1f));

                themesList.addView(card);
            }
        }
    }

    public void validate() {
        if(!validateVisibility()) {
            return;
        }

        validateThemesList();
    }

    private void validateThemesList() {
        LinearLayout themesList = findViewById(R.id.themesList);
        if(bundle != null) {
            for (int i = 0; i < themesList.getChildCount(); i++) {

                View child = themesList.getChildAt(i);
                if(child instanceof ThemeBundleImageView bundleCard) {
                    bundleCard.validate();
                }
            }

            setCreditCost();
        }
    }

    private boolean validateVisibility() {
        boolean isAvailable = bundle.getLockedItemCount() > 1;

        setVisibility(isAvailable ? VISIBLE : GONE);

        return isAvailable;
    }

    public void setCreditCost() {
        AppCompatTextView label_credits = findViewById(R.id.label_credits_cost);

        if(bundle == null) {
            return;
        }

        label_credits.setText(String.valueOf(bundle.getDiscountedBuyCredits()));
    }

    public void setBuyButtonListener(OnClickListener buyButtonListener) {
        View buyButton = findViewById(R.id.button_transactItem);

        if(buyButton != null) {
            buyButton.setOnClickListener(buyButtonListener);
        }
    }
}

File: app/src/main/java/com/TritiumGaming/phasmophobiaevidencepicker/activities/mainmenus/marketplace/views/items/ThemeBundleCardView.kt
package com.TritiumGaming.phasmophobiaevidencepicker.activities.mainmenus.marketplace.views.items

import android.content.Context
import android.util.AttributeSet
import android.view.ContextThemeWrapper
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import androidx.appcompat.widget.AppCompatTextView
import com.TritiumGaming.phasmophobiaevidencepicker.R
import com.TritiumGaming.phasmophobiaevidencepicker.data.viewmodels.models.firestore.theme.bundle.MarketThemeBundleModel
import com.TritiumGaming.phasmophobiaevidencepicker.utils.ColorUtils.getColorFromAttribute
import com.TritiumGaming.phasmophobiaevidencepicker.utils.ColorUtils.interpolate
import com.google.android.material.card.MaterialCardView

class ThemeBundleCardView : MaterialCardView {

    var bundle: MarketThemeBundleModel? = null
        set(value) {
            field = value

            field?.let { bundle ->
                val titleView = findViewById<AppCompatTextView>(R.id.label_bundleTitle)
                titleView?.text = bundle.name

                val costView = findViewById<AppCompatTextView>(R.id.label_credits_cost)
                costView?.text = bundle.discountedBuyCredits.toString()

                buildThemes()
            }
        }

    val creditCost: Long
        get() = bundle?.discountedBuyCredits ?: 0


    constructor(context: Context) : super(context, null)

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) :
            super(context, attrs, defStyleAttr)

    init {
        inflate(context, R.layout.item_marketplace_bundle, this)

        layoutParams = LinearLayout.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.WRAP_CONTENT)

        val strokeColor =
            getColorFromAttribute(context, R.attr.backgroundColorOnBackground)

        radius = 16f

        setStrokeColor(interpolate(resources.getColor(R.color.white), strokeColor, .25f))
        strokeWidth = 3

        useCompatPadding = true
        clipToPadding = false
    }

    private fun buildThemes() {
        val themesList = findViewById<LinearLayout>(R.id.themesList)

        bundle?.themes?.let { themes ->
            for (theme in themes) {
                val card = ThemeBundleImageView(
                    ContextThemeWrapper(context, theme.style), null, theme.style)
                card.setTheme(theme)

                card.layoutParams = LinearLayout.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.MATCH_PARENT, 1f)

                themesList.addView(card)
            }
        }
    }

    fun validate() {
        if (validateVisibility()) {
            validateThemesList()
        }
    }

    private fun validateThemesList() {
        val themesList = findViewById<LinearLayout>(R.id.themesList)
        bundle?.let {
            for (i in 0 until themesList.childCount) {
                val child = themesList.getChildAt(i)
                if (child is ThemeBundleImageView) { child.validate() }
            }

            setCreditCost()
        }
    }

    private fun validateVisibility(): Boolean {
        val isAvailable = bundle?.let { it.lockedItemCount > 1 } ?: false

        visibility = if (isAvailable) VISIBLE else GONE

        return isAvailable
    }

    private fun setCreditCost() {
        val creditsTextView = findViewById<AppCompatTextView>(R.id.label_credits_cost)

        bundle?.let { bundle -> creditsTextView.text = bundle.discountedBuyCredits.toString() }
    }

    fun setBuyButtonListener(buyButtonListener: OnClickListener?) {
        val buyButton = findViewById<View>(R.id.button_transactItem)

        buyButton?.setOnClickListener(buyButtonListener)
    }
}
